compiler/lib/global_deadcode.ml

```diff
@@ -0,0 +1,443 @@
+          add_cont_deps cont_h
+      | Poptrap cont -> add_cont_deps cont)
+    prog.blocks;
+  uses
+
+(** Return the set of variables used in a given expression *)
+let expr_vars e =
+  let vars = Var.Set.empty in
+  match e with
+  | Apply { f; args; _ } ->
+      let vars = Var.Set.add f vars in
+      List.fold_left ~f:(fun acc x -> Var.Set.add x acc) ~init:vars args
+  | Block (_, params, _) ->
+      Array.fold_left ~f:(fun acc x -> Var.Set.add x acc) ~init:vars params
+  | Field (z, _) -> Var.Set.add z vars
+  | Prim (_, args) ->
+      List.fold_left
+        ~f:(fun acc v ->
+          match v with
+          | Pv v -> Var.Set.add v vars
+          | Pc _ -> acc)
+        ~init:vars
+        args
+  (* We can ignore closures. We want the set of previously defined variables used
+     in the expression, so not parameters. The continuation may use some variables
+     but we will add these when we visit the body *)
+  | Constant _ | Closure (_, _) -> vars
+
+(** Compute the initial liveness of each variable in the program. 
+
+    A variable [x] is marked as [Top] if 
+    + It is used in an impure expression (as defined by [Pure_fun.pure_expr]);
+    + It is used in a conditonal/switch;
+    + It is raised by an exception;
+    + It is used in another stateful instruction (like setting a block or array field);
+    + Or, it is returned or applied to a function and the global flow analysis marked it as escaping.
+    
+    A variable [x[i]] is marked as [Live {i}] if it is used in an instruction where field [i] is referenced or set. *)
+let liveness prog pure_funs (global_info : Global_flow.info) =
+  let live_vars = Var.Tbl.make () Dead in
+  let add_top v = Var.Tbl.set live_vars v Top in
+  let add_live_field v i =
+    let live_fields =
+      match Var.Tbl.get live_vars v with
+      | Live fields -> Live (IntSet.add i fields)
+      | Top | Dead -> Live (IntSet.singleton i)
+    in
+    Var.Tbl.set live_vars v live_fields
+  in
+  let live_instruction i =
+    match i with
+    (* If e is impure, set all variables in e as Top. The only exception is for function applications,
+       where we may be able to do better. Global flow gives us information about which arguments in
+       a function application escape, so set only these as top. *)
+    | Let (_, e) -> (
+        if not (Pure_fun.pure_expr pure_funs e)
+        then
+          match e with
+          | Apply { f; args; _ } ->
+              add_top f;
+              List.iter
+                ~f:(fun x -> if variable_may_escape x global_info then add_top x)
+                args
+          | Block (_, _, _) | Field (_, _) | Closure (_, _) | Constant _ | Prim (_, _) ->
+              let vars = expr_vars e in
+              Var.Set.iter add_top vars)
+    | Set_field (x, i, y) ->
+        add_live_field x i;
+        add_top y
+    | Array_set (x, y, z) ->
+        add_top x;
+        add_top y;
+        add_top z
+    | Offset_ref (x, i) -> add_live_field x i
```

Comment:
`Offset_ref (x, i)` adds `i` to the field `0` of the value of `x`
```suggestion
    | Offset_ref (x, _) -> add_live_field x 0
```

Reply:
Okay I'm not getting printf failures with this change! I never paid much attention to this expression...what is it?

Reply:
[Fix offset_ref liveness](https://github.com/ocsigen/js_of_ocaml/pull/1503/commits/1c492f09dd150a815d059b85d8c76d41d467cf85)

Reply:
This is a shortcut for `x := !x + i`, corresponding to the `OFFSETREF` bytecode instruction. I'm not sure we really need it, but well...